logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamImpl.java

```diff
@@ -193,78 +174,78 @@ public void removeOnCommitPositionUpdatedCondition(final ActorCondition conditio
-    if (actor.isClosed()) {
-      return CompletableActorFuture.completedExceptionally(new RuntimeException("Actor is closed"));
+  protected void onActorClosing() {
+    LOG.info("On closing logstream {} close {} readers", logName, readers.size());
+    readers.forEach(LogStreamReader::close);
+    LOG.info("Close log storage with name {}", logName);
+    logStorage.close();
+  }
+
+  @Override
+  protected void onActorClosed() {
+    if (closeError != null) {
+      closeFuture.completeExceptionally(closeError);
+    } else {
+      closeFuture.complete(null);
     }
+  }
 
-    final var writerFuture = new CompletableActorFuture<LogStreamRecordWriter>();
-    actor.run(() -> createWriter(writerFuture, LogStreamWriterImpl::new));
-    return writerFuture;
+  @Override
+  public void close() {
+    closeAsync().join();
   }
 
   @Override
-  public ActorFuture<LogStreamBatchWriter> newLogStreamBatchWriter() {
-    // this should be replaced after refactoring the actor control
+  public ActorFuture<Void> closeAsync() {
     if (actor.isClosed()) {
-      return CompletableActorFuture.completedExceptionally(new RuntimeException("Actor is closed"));
+      return closeFuture;
     }
 
-    final var writerFuture = new CompletableActorFuture<LogStreamBatchWriter>();
-    actor.run(() -> createWriter(writerFuture, LogStreamBatchWriterImpl::new));
-    return writerFuture;
+    actor.call(
+        () ->
+            closeAppender()
+                .onComplete(
+                    (nothing, appenderError) -> {
+                      closeError = appenderError;
+                      actor.close();
+                    }));
+    return closeFuture;
+  }
+
+  private void internalSetCommitPosition(final long commitPosition) {
+    if (commitPosition > this.commitPosition) {
+      this.commitPosition = commitPosition;
+      onCommitPositionUpdatedConditions.signalConsumers();
+    }
   }
 
   private <T extends LogStreamWriter> void createWriter(
       final CompletableActorFuture<T> writerFuture, final WriterCreator<T> creator) {
-    final var alreadyOpenWriters = openWriterCount.getAndIncrement();
-    if (alreadyOpenWriters == 0) {
-      openAppender().onComplete(onOpenAppender(writerFuture, creator));
-    } else if (appender != null) {
-      writerFuture.complete(creator.create(partitionId, writeBuffer, this::releaseWriter));
+    if (appender != null) {
+      writerFuture.complete(creator.create(partitionId, writeBuffer));
     } else if (appenderFuture != null) {
       appenderFuture.onComplete(onOpenAppender(writerFuture, creator));
     } else {
-      final var errorMsg =
-          String.format(
-              "Expected to have an open appender, since we have already %d open writers",
-              alreadyOpenWriters);
-      writerFuture.completeExceptionally(new IllegalStateException(errorMsg));
+      openAppender().onComplete(onOpenAppender(writerFuture, creator));
```

Comment:
So this means the first writer will create the appender right?

Reply:
Yes.